src/client/auth.js

```diff
@@ -14,9 +15,56 @@ function validateOptions (options) {
   }
 }
 
+function retry (fn, maxRetries, interval = 1000) {
+  return new Promise((resolve, reject) => {
+    let retries = 0
+    const retryFn = () => {
+      retries++
+      fn().then(resolve).catch(err => {
+        debug('retry', retries, err)
+        if (retries >= maxRetries) {
+          reject(err)
+        } else {
+          setTimeout(retryFn, interval)
+        }
+      })
+    }
+    retryFn()
+  })
+}
+
 async function realmAuthenticate (options) {
   validateOptions(options)
-  throw new Error('Not implemented')
+
+  options.authflow = new PrismarineAuth(options.username, options.profilesFolder, options, options.onMsaCode)
+
+  const api = RealmAPI.from(options.authflow, 'bedrock')
+  const realms = await api.getRealms()
+
+  debug('realms', realms)
+
+  if (!realms || !realms.length) throw Error('Couldn\'t find any Realms for the authenticated account')
+
+  let realm
+
+  if (options.realms.realmId) {
+    realm = realms.find(e => e.id === Number(options.realms.realmId))
+  } else if (options.realms.realmInvite) {
+    realm = await api.getRealmFromInvite(options.realms.realmInvite)
+  } else if (options.realms.pickRealm) {
+    if (typeof options.realms.pickRealm !== 'function') throw Error('realms.pickRealm must be a function')
+    realm = await options.realms.pickRealm(realms)
+  }
+
+  if (!realm) throw Error('Couldn\'t find a Realm to connect to. Authenticated account must be the owner or has been invited to the Realm.')
+
+  const { address } = await retry(async () => realm.getAddress(), 5)
+  const [host, port] = address.split(':')
```

Comment:
This should probably be handled in prismarine-realms

Reply:
Return an array with `[host, port]` or an object `{ host: '', port: '' }` ?

Reply:
Either way works, if later on they move to IPv6 addresses for example the ':' splitting mechanics will break in user code